### ISSUE TITLE
Filter object properties by type

### ISSUE THREAD

**Issue by turolla** [microsoft/TypeScript#38646]
**TypeScript Version:**  3.9.2

**Search Terms:**
mapped type filter object properties never 

**Expected behavior:**
The type
`type FilteredKeys<T, U> = { [P in keyof T]: T[P] extends U ? P : never }[keyof T];`
that allows to extract object properties by type should work everywhere.

**Actual behavior:**
`type FilteredKeys<T, U> = { [P in keyof T]: T[P] extends U ? P : never }[keyof T];`
only works externally of the declaring type

**Related Issues:**
#23199 

**Code**
```ts
type FilteredKeys<T, U> = { [P in keyof T]: T[P] extends U ? P : never }[keyof T];

type FooItem<T> = T extends number ? { value: T } : undefined;

type FooMap<T> = {
    // [key in keyof T]: FooItem<T[key]>  // <-- This works
    [key in FilteredKeys<T, number>]: FooItem<T[key]>  // <-- This doesn't work
}

class Bar {
    a = 1;
    b = 2;
    c = '';    

    doSomething() {
        const fooMap: FooMap<this> = undefined!; // Omitted...
        const a = fooMap.a.value; // <-- No error expected
        const c = fooMap.c.value; // <-- Error expected
    }
}

const fooMap2: FooMap<Bar> = undefined!; // Omitted...
const a = fooMap2.a.value; // <-- No error expected
const c = fooMap2.c.value; // <-- Error expected

```

<details><summary><b>Output</b></summary>

```ts
"use strict";
class Bar {
    constructor() {
        this.a = 1;
        this.b = 2;
        this.c = '';
    }
    doSomething() {
        const fooMap = undefined; // Omitted...
        const a = fooMap.a.value; // <-- No error expected
        const c = fooMap.c.value; // <-- Error expected
    }
}
const fooMap2 = undefined; // Omitted...
const a = fooMap2.a.value; // <-- No error expected
const c = fooMap2.c.value; // <-- Error expected

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/index.html#code/C4TwDgpgBAYglgG2BAThAJgaQiAzgHgBUAaKAVQD4oBeKAbygG0AFKOAOygGscB7AMyiEAugC4hLYVAgAPZO3S5yUAPxRW49hABuqKAF9GPEAKHCA3AFgAUDdCRYvXgElkAWyJVahaXIgKldgBXNwAjPTUGbQBDBCCIcR99KHEghQh+DgwrW2t7aBgnAFlosE8aehsoaqgAelqmYzZOY1MRcUKXdyIjHGEqOob8AFphoQALOCUAd14ULlwqmt6QZthEZDQsHAISKGCw1AoxRy6ID0IV-ur6qBGxwkmldF4IXHYAcmAoWfmbfRsNgAxghorglAAhaIoSrWGpQaIVACMOXhoQqACZUTUgRUPh9zPDAXCai8AMq8NwQYCTdgAcwAFABKWHw+FA3jsXDffjFUodPllGlTLxQNLoDJZdAAQkJtwA8m44MBkOgAHQapZsqAcrnfRG0Xm8EpgNXRNUxOIQOVDUZQAByvGkKBQc18kCBqq1bN13J1FSNJrVQItsXiNrudoAoi63bIPV6SdUAdYU8DOX7A6UMQLjaV8FCUKLxZKtDKI4rlaqNWr03qEQHBRizaGrRH7g6nahXTD4xBPRg637cYam8HW+HBpGxjGe+7+4mgA)

**Comment by RyanCavanaugh** [comment]
@weswigham thoughts?

**Comment by dumasss163** [comment]
you could do this to make it compile:

```ts
type FilteredKeys<T, U> = keyof { [P in keyof T]: T[P] extends U ? P : never };

```

But this still has other problem sometimes, see: 
[Property Type inference problem when picking property with index key type · Issue #39945 · microsoft/TypeScript[2.1k,62.9k:62.9k,8.4k]](https://github.com/microsoft/TypeScript/issues/39945)

**Comment by shauns** [comment]
In 4.1 (beta) you can do something like:

```
type PickByValueType<T, U> = {
  [K in keyof T as T[K] extends U ? K : never]: T[K]
}
```

**Comment by turolla** [comment]
Key remapping  is a very nice and useful feature.
Unfortunately the issue seems to be still here 

https://www.typescriptlang.org/play?ts=4.1.0-dev.20200929#code/C4TwDgpgBAYg9nAksCBbAPAFQHxQLxSZQQAeKAdgCYDOU5ArqgEYQBOUA-FAN5QBuAQwA29CAC5CUAL5QJ9KhABmAS3IRKAbgCwAKF2hIsBAFkBYLLgLddUW1AD09qAG0A1hBBRVUdyDiLCAF0JeCQUDEw3D0DcByd0AFoEwgALZVoAdzhWV2obO2cAaS9yHw9-SQFaSMLA4jIIKloGZjZOKGKJNT42YKMwtCwimLioROTMNNpKOAhqcgByYCgsnN0pXV0AYyEq2gAhAXZrHTsoAXwoAEZtU7smS4AmW7Oty4WF2-zbGYBlOFQEGAaXIAHMABQASh43zOUC2cHI1GWihMZhCaPMwPSlig8koSlU6gAhBpRgB5VDKYAoSgAOgZsNeiOR50uqLgpjAdIEdMEIggZMcYySUAAcnBiKxWNl6pAtrSmXYEUjlm8CByuXStnzhKIhfFRQBRaWy0jyxV3WwbHQ27YslGYx4YzlmdCHVi4-GEtSUUkUqk09QMun21VsjVOnm6gUGkXJCVSmXsc0QBXqMOs9VQTVmR7amP60bjKAm5NytOWoA

**Comment by alvis** [comment]
In 4.1, as @shauns mentioned, you can filter out object keys which don't fit the type you want. However, unfortunately, it still can't give you a list of keys which comply with the desired type by the conventional `key of` trick. i.e. The following doesn't work:

```ts
type SelectKeysByValueType<T, U> = {
  [K in keyof T as T[K] extends U ? K : never]: T[K]
}[keyof T] // <~ Type 'keyof T' cannot be used to index type
```

**Comment by ziofat** [comment]
Any updates on this issue?

I have a similar issue here:

https://www.typescriptlang.org/play?ts=4.2.3#code/C4TwDgpgBAcg9gOwGIFcEGNgEtEEMA2APACoB8UAvFAN4CwAUFE1ANoDSUWCUA1hCHABmUYlFwBnEewC6UCAA9gEBABNJqDNkRQA-FAQQAbhABOUAFxQ20ywFF56fChUQSMgDRQNmHAlIBuBgBfBgYuJRNBXHRoAGFEQSwAcxoGZn1cAFsIS3FgEy4kwMZmdAJ8ACNonksACgBKSnJDOCwVYvTMrHlLCrg4fAhcbgAfKFqGpqg+gaGEeuKQ+gZHCUkAWWHcJNNCWLlFZTUoeIREpPI6EqZ0RDyTFEw4E1rbs+TLeGQ0HzwiWNIjSu6XSb3EswAdPg4ElXglkhCEFkIAs0iCbndIdDYW9zhCyvhKtVUdcQWCsTC4e8khCuvISeklksGEA

**Comment by pascalgn** [comment]
As a workaround, I think this could work:
```ts
type OriginalType = {
    str: string;
    num: number;
}

type Values<T> = T[keyof T];

type PickKeys<T, U> = Values<{
  [K in keyof T]: T[K] extends U ? K : never;
}>;

type NumberKeys = PickKeys<OriginalType, number>;
type NumberType = Pick<OriginalType, NumberKeys>;

const ok: NumberType = { num: 1 }; // ok
const err: NumberType = { num: 1, str: "" }; // error
```
With the original example rewritten:
```ts
type FooMap<T> = {
    [key in PickKeys<T, number>]: T[key]
}

const ok1: FooMap<OriginalType> = { num: 1 }; // ok
const err2: FooMap<OriginalType> = { num: 1, str: "" }; // error
```

**Comment by aspic-fish** [comment]
I have a workaround, just infer type before mapping it. 
So rewrite this:
```javascript
type FooMap<T> = {
    [key in FilteredKeys<T, number>]: FooItem<T[key]> 
}
```
to this:
```javascript
type FooMap<T> = T extends infer TT 
  ?  {
    [key in FilteredKeys<T, number>]: FooItem<T[key]>
  }
  : never
```
 [turolla case](https://www.typescriptlang.org/play?ts=4.5.4#code/C4TwDgpgBAYglgG2BAThAJgaQiAzgHgBUAaKAVQD4oBeKAbygG0AFKOAOygGscB7AMyiEAugC4hLYVAgAPZO3S5yUAPxRW49hABuqKAF9GPEAKHCA3AFgAUDdCRYvXgElkAWyJVahaXIgKldgBXNwAjPTUGbQBDBCCIcR99KHEghQh+DgwrW2t7aBgnAFlosE8aIV95RTZ2fj1CHxsoFtV6ZtaAek6mY1ruPkERcUKXdyIjHGEqKG6ofABaBaEACzglAHdeFC5cDpbJkH74JFQMbDwiUmCw1AoxRzGID0JD6Za5xeXCNaV0XgguHYAHJgFAtjt9gYoZodKgbDYAMYIaK4JQAIWiKHa1laUGiFQAjDk8aEKgAmEmtREVYHA8x4hG41r-ADKvDcEGAa3YAHMABQAShxeLxiN47FwYP4xVKI1lZW56y8UDS6AyWXQAEIGXMAPJuODAZDoAB05qhYolUvxFRlvBKYFN0VNMTiEF1PS+UAAcrxpCgUNtfJBESbLdTrWCabR7Y7TYjXbF4p75ksoABRQPB2Sh8PMlr6GxF3LiyXShXk+UO0r4TEoFVqjVabWpg1Gk3m01IqO22OV51J92p71+gNB7G5iBhjA98tQGNQOOlckJocp2Ze9NZich6f5oA)

[ziofat case](https://www.typescriptlang.org/play?ts=4.2.3#code/C4TwDgpgBAcg9gOwGIFcEGNgEtEEMA2APACoB8UAvFMVBAB7AQIAmAzlFggGYQBO1xALAAoKGKgB+KAG8R4+VADaAaQ4IoAawgg4XalFztiKgLq0GTNlFQZsiSVAQQAbnygAuKMpOeAonXR8FGYIElMAGms0TBwEUgBuOXEAXySxTydXXhEc4U5GXi5cdGgAYUQuLABzGTTHXABbCE9WYF5OKsTRcXQCfAAjYo1PAAoASkpyZzgsZi75Bqw6T364OHwIXHUAHygR8cmoVfXNhDGu1OERQMN2AFkt3Cq+QlLzRhZ2cu5q8llusToRCtXgoTBwXgjIE-KqeeDIaJ2BAEV6kCb-BQ9YEnAB0+DgVShFWqOORTXOdXk0NYuPxhOhlSqON6+AGQwpAIU1NpBKJMJxizoHPkl0uuREQA)

**Comment by turolla** [comment]
> I have a workaround, just infer type before mapping it. So rewrite this:
> 
> ```js
> type FooMap<T> = {
>     [key in FilteredKeys<T, number>]: FooItem<T[key]> 
> }
> ```
> 
> to this:
> 
> ```js
> type FooMap<T> = T extends infer TT 
>   ?  {
>     [key in FilteredKeys<T, number>]: FooItem<T[key]>
>   }
>   : never
> ```
> 
> [turolla case](https://www.typescriptlang.org/play?ts=4.5.4#code/C4TwDgpgBAYglgG2BAThAJgaQiAzgHgBUAaKAVQD4oBeKAbygG0AFKOAOygGscB7AMyiEAugC4hLYVAgAPZO3S5yUAPxRW49hABuqKAF9GPEAKHCA3AFgAUDdCRYvXgElkAWyJVahaXIgKldgBXNwAjPTUGbQBDBCCIcR99KHEghQh+DgwrW2t7aBgnAFlosE8aIV95RTZ2fj1CHxsoFtV6ZtaAek6mY1ruPkERcUKXdyIjHGEqKG6ofABaBaEACzglAHdeFC5cDpbJkH74JFQMbDwiUmCw1AoxRzGID0JD6Za5xeXCNaV0XgguHYAHJgFAtjt9gYoZodKgbDYAMYIaK4JQAIWiKHa1laUGiFQAjDk8aEKgAmEmtREVYHA8x4hG41r-ADKvDcEGAa3YAHMABQAShxeLxiN47FwYP4xVKI1lZW56y8UDS6AyWXQAEIGXMAPJuODAZDoAB05qhYolUvxFRlvBKYFN0VNMTiEF1PS+UAAcrxpCgUNtfJBESbLdTrWCabR7Y7TYjXbF4p75ksoABRQPB2Sh8PMlr6GxF3LiyXShXk+UO0r4TEoFVqjVabWpg1Gk3m01IqO22OV51J92p71+gNB7G5iBhjA98tQGNQOOlckJocp2Ze9NZich6f5oA)
> 
> [ziofat case](https://www.typescriptlang.org/play?ts=4.2.3#code/C4TwDgpgBAcg9gOwGIFcEGNgEtEEMA2APACoB8UAvFMVBAB7AQIAmAzlFggGYQBO1xALAAoKGKgB+KAG8R4+VADaAaQ4IoAawgg4XalFztiKgLq0GTNlFQZsiSVAQQAbnygAuKMpOeAonXR8FGYIElMAGms0TBwEUgBuOXEAXySxTydXXhEc4U5GXi5cdGgAYUQuLABzGTTHXABbCE9WYF5OKsTRcXQCfAAjYo1PAAoASkpyZzgsZi75Bqw6T364OHwIXHUAHygR8cmoVfXNhDGu1OERQMN2AFkt3Cq+QlLzRhZ2cu5q8llusToRCtXgoTBwXgjIE-KqeeDIaJ2BAEV6kCb-BQ9YEnAB0+DgVShFWqOORTXOdXk0NYuPxhOhlSqON6+AGQwpAIU1NpBKJMJxizoHPkl0uuREQA)

@aspic-fish YOU ARE A GENIUS !!!! Thank you so much!!!

**Comment by edvald** [comment]
Here's a solution we found to work for us (in this case omitting as opposed to picking), including a recursive option:

```typescript
type OmitKeysByValueType<T, U> = { [P in keyof T]: T[P] extends U ? never : P }[keyof T]

type OmitByValueType<T, V> = T extends infer _
  ? {
      [key in OmitKeysByValueType<T, V>]: T[key]
    }
  : never

type FunctionsOmitted = OmitByValueType<A, Function>

type OmitByValueTypeRecursive<T, V> = T extends object
  ? T extends infer _
    ? {
        [key in OmitKeysByValueType<T, V>]: OmitByValueTypeRecursive<T[key], V>
      }
    : never
  : T
```

Usage:

```typescript
interface A {
  someFunctionIDoNotWant: () => void // <- get rid of this
  somePropIDoWant: number
}

interface B {
  prop: string
  prop2: A
}

type AMinusFunctions = OmitByValueType<A, Function>
// -> { somePropIDoWant: true }

type BMinusFunctionsRecursively = OmitByValueTypeRecursive<B, Function>
// -> { prop: string; prop2: {  somePropIDoWant: true } }
```

cc @Orzelius

**Comment by RyanCavanaugh** [comment]
A few observations:
 * The nonworking form isn't something we can analyze for an arbitrary type parameter like `this` and get the right answer in the general case
 * The working form... works?
 * Looks like there are other workarounds too

The real missing feature is #48992 to handle this in a more analyzable form, so I think we can treat this as a duplicate of that one
